File: python/ReorderList143.py
# You are given the head of a singly linked-list. The list can be represented as:
# L0 → L1 → … → Ln - 1 → Ln
# Reorder the list to be on the following form:
# L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …
# You may not modify the values in the list's nodes. Only nodes themselves may be changed.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reorderList(head: Optional[ListNode]) -> None:
    if not head or not head.next:
        return
    slow, fast = head, head
    while fast.next and fast.next.next:
        slow, fast = slow.next, fast.next.next
    end = None
    while slow:
        slow.next, end, slow = end, slow, slow.next
    while head and end:
        head.next, end.next, head, end = end, head.next, head.next, end.next


if __name__ == "__main__":
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(4)
    head_outer.next.next.next.next = ListNode(5)
    reorderList(head_outer)
    while head_outer:
        print(head_outer.val, end=" -> ")
        head_outer = head_outer.next
    print()
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(4)
    head_outer.next.next.next.next = ListNode(5)
    head_outer.next.next.next.next.next = ListNode(6)
    reorderList(head_outer)
    while head_outer:
        print(head_outer.val, end=" -> ")
        head_outer = head_outer.next
    print()
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(4)
    head_outer.next.next.next.next = ListNode(5)
    head_outer.next.next.next.next.next = ListNode(6)
    head_outer.next.next.next.next.next.next = ListNode(7)
    reorderList(head_outer)
    while head_outer:
        print(head_outer.val, end=" -> ")
        head_outer = head_outer.next
    print()
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(4)
    head_outer.next.next.next.next = ListNode(5)
    head_outer.next.next.next.next.next = ListNode(6)
    head_outer.next.next.next.next.next.next = ListNode(7)
    head_outer.next.next.next.next.next.next.next = ListNode(8)
    reorderList(head_outer)
    while head_outer:
        print(head_outer.val, end=" -> ")
        head_outer = head_outer.next

File: java/ReorderList143.java
/*
You are given the head of a singly linked-list. The list can be represented as:
L0 → L1 → … → Ln - 1 → Ln
Reorder the list to be on the following form:
L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …
You may not modify the values in the list's nodes. Only nodes themselves may be changed.
*/
public class ReorderList143 {
    public static void main(String[] args) {
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        reorderList(head);
        while (head != null) {
            System.out.print(head.val + " ");
            head = head.next;
        }
        System.out.println();

        head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
        reorderList(head);
        while (head != null) {
            System.out.print(head.val + " ");
            head = head.next;
        }
    }


    public static void reorderList(ListNode head) {
        if (head == null || head.next == null) return;
        ListNode slow = head, fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode end = null, next;
        while (slow != null) {
            next = slow.next;
            slow.next = end;
            end = slow;
            slow = next;
        }
        while (head != null && end != null) {
            ListNode nextHead = head.next, nextEnd = end.next;
            head.next = end;
            end.next = nextHead;
            head = nextHead;
            end = nextEnd;
        }
    }

    public static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}
